src/libxmljs.cc

```diff
@@ -9,39 +9,78 @@
 #include "xml_node.h"
 #include "xml_sax_parser.h"
 
-namespace libxmljs {
+#if NAUV_UVVERSION < 0xb0b
+#include "uvcompat.h"
+#endif
 
-// ensure destruction at exit time
-// v8 doesn't cleanup its resources
-LibXMLJS LibXMLJS::instance;
+namespace libxmljs {
 
-// track how much memory libxml2 is using
-int xml_memory_used = 0;
+bool tlsInitialized = false;
+uv_key_t tlsKey;
+bool isAsync = false; // Only set on V8 thread when no workers are running
+int numWorkers = 0; // Only access from V8 thread
 
-// wrapper for xmlMemMalloc to update v8's knowledge of memory used
-// the GC relies on this information
-void* xmlMemMallocWrap(size_t size)
-{
-    void* res = xmlMemMalloc(size);
+// Set up in V8 thread
+WorkerParent::WorkerParent() : memAdjustments(0) {
+    if (!tlsInitialized)
+    {
+        uv_key_create(&tlsKey);
+        tlsInitialized = true;
+    }
+    if (numWorkers++ == 0)
+    {
+        isAsync = true;
+    }
+}
 
-    // no need to udpate memory if we didn't allocate
-    if (!res)
+// Tear down in V8 thread
+WorkerParent::~WorkerParent() {
+    NanAdjustExternalMemory(memAdjustments);
+    if (--numWorkers == 0)
     {
-        return res;
+        isAsync = false;
     }
+}
 
-    const int diff = xmlMemUsed() - xml_memory_used;
-    xml_memory_used += diff;
-    NanAdjustExternalMemory(diff);
-    return res;
+// Set up in worker thread
+WorkerSentinel::WorkerSentinel(WorkerParent& parent) : parent(parent) {
+    uv_key_set(&tlsKey, this);
 }
 
-// wrapper for xmlMemFree to update v8's knowledge of memory used
-// the GC relies on this information
-void xmlMemFreeWrap(void* p)
-{
-    xmlMemFree(p);
+// Tear down in worker thread
+WorkerSentinel::~WorkerSentinel() {
+    uv_key_set(&tlsKey, NULL);
+}
+
+struct memHdr {
+    size_t size;
+};
+
+// When we move to C++11 one day, the following would become constexpr,
+// and SLIGN_SIZE could perhaps become sizeof(std::max_align_t).
+#define ALIGN_SIZE sizeof(double)
+#define HDR_SIZE (((sizeof(memHdr) + (ALIGN_SIZE - 1))/ALIGN_SIZE)*ALIGN_SIZE)
```

Comment:
This is based on [what libxml does](https://github.com/GNOME/libxml2/blob/38bbd3412d1c2a2016cf5a1a221ad4010216d28e/xmlmemory.c#L109). I assumed that since the debug memory code used to work so far, that approach should be fine, while everything more advanced has some chance that some compiler might balk at it. But I'll be happy with some more recent and legible way to express this if I'm not held responsible for any compatibility issues that might cause…